src/Dependencies/PooledObjects/PooledDelegates.cs

```diff
@@ -0,0 +1,404 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.CodeAnalysis.PooledObjects
+{
+    /// <summary>
+    /// Provides pooled delegate instances to help avoid closure allocations for delegates that require a state argument
+    /// with APIs that do not provide appropriate overloads with state arguments.
+    /// </summary>
+    internal static class PooledDelegates
+    {
+        private static class DefaultDelegatePool<T>
+            where T : class, new()
+        {
+            public static readonly ObjectPool<T> Instance = new ObjectPool<T>(() => new T(), 20);
+        }
+
+        private static Releaser GetPooledDelegate<TPooled, TArg, TUnboundDelegate, TBoundDelegate>(TUnboundDelegate unboundDelegate, TArg argument, out TBoundDelegate boundDelegate)
+            where TPooled : AbstractDelegateWithBoundArgument<TArg, TUnboundDelegate, TBoundDelegate>, new()
+            where TUnboundDelegate : Delegate
+            where TBoundDelegate : Delegate
+        {
+            var obj = DefaultDelegatePool<TPooled>.Instance.Allocate();
+            obj.Argument = argument;
+            obj.UnboundDelegate = unboundDelegate;
+            boundDelegate = obj.BoundDelegate;
+
+            return new Releaser(
+                obj,
+                obj =>
+                {
+                    var pooledObject = (TPooled)obj;
+                    pooledObject.Clear();
+                    DefaultDelegatePool<TPooled>.Instance.Free(pooledObject);
+                });
+        }
+
+        /// <summary>
+        /// Gets an <see cref="Action"/> delegate, which calls <paramref name="unboundAction"/> with the specified
+        /// <paramref name="argument"/>. The resulting <paramref name="boundAction"/> may be called any number of times
+        /// until the returned <see cref="Releaser"/> is disposed.
+        /// </summary>
+        /// <example>
+        /// <para>The following example shows the use of a capturing delegate for a callback action that requires an
+        /// argument:</para>
+        ///
+        /// <code>
+        /// int x = 3;
+        /// RunWithActionCallback(() => this.DoSomething(x));
+        /// </code>
+        ///
+        /// <para>The following example shows the use of a pooled delegate to avoid capturing allocations for the same
+        /// callback action:</para>
+        ///
+        /// <code>
+        /// int x = 3;
+        /// using var _ = GetPooledAction(arg => arg.self.DoSomething(arg.x), (self: this, x), out Action action);
```

Comment:
>using var _ = GetPooledAction(arg => arg.self.DoSomething(arg.x), (self: this, x), out Action action); [](start = 12, length = 102)

This delegate is not fully equivalent to the original one (x is captured by value at the point where the delegate is created). I think this should be clearly documented, with clear explanation of what the difference is.

Reply:
The same limitation would apply to any case where an overload is provided for the arguments. I'll clarify that the new behavior is equivalent to that case, with the caveat that the behavior of `boundAction` after the releaser is disposed is undefined.

Reply:
At this point, the passing of user-defined state arguments to a method that takes a delegate is a well-established pattern with consistent semantics. Since I have not deviated from the well-known semantics, I don't believe further elaboration on the wider pattern is necessary in such a limited scope. I verified that the releaser semantics have been documented which should suffice for this API.